### PR TITLE
Fixes 04

### PR DESCRIPTION
**Pull request by alesstimec**

### CODE REVIEW

File: simulation/simulation.go [hunk 1]
```diff
@@ -114,27 +110,41 @@ func (e *entitySet) create(ctx context.Context, sim *Simulation) {
...
-		if err != nil {
-			sim.error(errors.Trace(err))
-			return
-		}
-		createEntity(ctx, cfg, copyAttributes(e.attributes), sim)
-	}
+	e.numberOfEntities = numberOfEntities
+	e.numberOfCreatedEntities = 0
+
+	// add a job that will all all the entities
```
Comment (cmars) [inline simulation/simulation.go]: add all

File: simulation/simulation.go [hunk 2]
```diff
@@ -114,27 +110,41 @@ func (e *entitySet) create(ctx context.Context, sim *Simulation) {
...
+		e.createEntity(ctx, sim)
+	})
+
 	return
 }
 
+func (e *entitySet) createEntity(ctx context.Context, sim *Simulation) {
+	e.mu.Lock()
+	if e.numberOfCreatedEntities >= e.numberOfEntities {
+		return
```
Comment (cmars) [inline simulation/simulation.go:137]: Lock would still be held here. I think you'd want to move the defer unlock up, to immediately follow the lock.

File: simulation/executor.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+// addJob will create a new timer that will fire after the delay period. Once
+// the timer fires, the job is put on the execution queue.
+func (e *executor) addJob(ctx context.Context, delay time.Duration, j job) {
+	go func() {
+		timer := time.NewTimer(delay)
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
```
Comment (cmars) [inline simulation/executor.go]: Or `time.After(delay)`?

**Review by cmars — APPROVED** [review]
LGTM with deadlock issue addressed.

File: simulation/simulation.go (hunk 1 above)
Comment (alesstimec) [reply simulation/simulation.go]: fixed

File: simulation/simulation.go (hunk 2 above)
Comment (alesstimec) [reply simulation/simulation.go:137]: fixed

File: simulation/executor.go (hunk 3 above)
Comment (alesstimec) [reply simulation/executor.go]: done

**Pull request merged**